tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,652 @@
+	delayedProposalsList map[types.EpochID][][]types.ATXID
+
+	// TODO(nkryuchkov): consider removing
+	lateProposalsMu   sync.RWMutex
+	lateProposalsList map[types.EpochID][][]types.ATXID
+
+	votesMu      sync.RWMutex
+	votesFor     map[EpochRoundPair][]types.Hash32
+	votesAgainst map[EpochRoundPair][]types.Hash32
+
+	beaconsMu sync.RWMutex
+	beacons   map[types.EpochID]types.Hash32
+	// beaconsReady indicates if beacons are ready.
+	// If a beacon for an epoch becomes ready, channel for this epoch becomes closed.
+	beaconsReady map[types.EpochID]chan struct{}
+
+	backgroundWG sync.WaitGroup
+}
+
+func New(
+	conf Config,
+	messageReceiver messageReceiver,
+	messageSender messageSender,
+	beaconCalculator beaconCalculator,
+	epochATXGetter epochATXGetter,
+	tortoiseBeaconDB tortoiseBeaconDB,
+	weakCoin weakCoin,
+	layerTicker chan types.LayerID,
+	logger log.Log,
+) *TortoiseBeacon {
+	return &TortoiseBeacon{
+		Log:                  logger,
+		Closer:               NewCloser(),
+		config:               conf,
+		messageReceiver:      messageReceiver,
+		messageSender:        messageSender,
+		beaconCalculator:     beaconCalculator,
+		epochATXGetter:       epochATXGetter,
+		tortoiseBeaconDB:     tortoiseBeaconDB,
+		weakCoin:             weakCoin,
+		layerTicker:          layerTicker,
+		networkDelta:         time.Duration(conf.WakeupDelta) * time.Second,
+		currentRounds:        make(map[types.EpochID]int),
+		timelyProposalsMap:   make(map[types.Hash32][]types.ATXID),
+		timelyProposalsList:  make(map[types.EpochID][][]types.ATXID),
+		delayedProposalsMap:  make(map[types.Hash32][]types.ATXID),
+		delayedProposalsList: make(map[types.EpochID][][]types.ATXID),
+		lateProposalsList:    make(map[types.EpochID][][]types.ATXID),
+		votesFor:             make(map[EpochRoundPair][]types.Hash32),
+		votesAgainst:         make(map[EpochRoundPair][]types.Hash32),
+		beacons:              make(map[types.EpochID]types.Hash32),
+		beaconsReady:         make(map[types.EpochID]chan struct{}),
+	}
+}
+
+// Start starts listening for layers and outputs.
+func (tb *TortoiseBeacon) Start() error {
+	tb.Log.Info("Starting %v", protoName)
+
+	tb.backgroundWG.Add(1)
+	go func() {
+		defer tb.backgroundWG.Done()
+
+		tb.listenLayers()
+	}()
+
+	tb.backgroundWG.Add(1)
+	go func() {
+		defer tb.backgroundWG.Done()
+
+		if err := tb.listenMessages(); err != nil {
+			tb.Log.With().Error("Error while listening for messages: %v", log.Err(err))
+			return
+		}
+	}()
+
+	tb.backgroundWG.Add(1)
+	go func() {
+		defer tb.backgroundWG.Done()
+
+		tb.cleanupLoop()
+	}()
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) Close() error {
+	tb.Log.Info("Closing %v", protoName)
+	tb.Closer.Close()
+	tb.backgroundWG.Wait() // Wait until background goroutines finish
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) Get(epochID types.EpochID) (types.Hash32, error) {
+	if tb.tortoiseBeaconDB != nil {
```

Comment:
why would this be nil? isn't it initialised in constructor?

Reply:
It is, but `nil` may be passed to the constructor. In this case, DB is not used. For example, we can use `nil` in tests, where we need to test parts that don't require DB.